doc/release/upcoming_changes/25168.change.rst

```diff
@@ -7,3 +7,15 @@ Now `numpy.array` and `numpy.asarray` support three values for ``copy`` paramete
 * ``False`` - Never make a copy. If a copy is required a ``ValueError`` is raised.
 
 The meaning of ``False`` changed as it now raises an exception if a copy is needed.
+
+The ``__array__`` special method now takes a ``copy`` keyword argument.
+-----------------------------------------------------------------------
+
+NumPy will pass ``copy`` to the ``__array__`` special method in situations where
+it would be set to a non-default value (e.g. in a call to
+``np.asarray(some_object, copy=False)``). Currently, if an
+unexpected keyword argument error is raised after this, NumPy will print a
+warning and re-try without the ``copy`` keyword argument. Implementations of
+objects implementing the ``__array__`` protocol should accept a ``copy`` keyword
+argument with the same meaning as when passed to `numpy.array` or
+`numpy.asarray`.
```

Comment:
Happy to move it to a new issue, but thinking through how to update some `__array__` implementations, I am wondering: what are the expectations from numpy how an `__array__` implementations handles this keyword fully or partially? 

When `copy=True` is being passed, does numpy assume that the `__array__` implementation already made that copy? But what if the implementation defers dtype casting to numpy (which is currently possible), then it might be a waste to already copy up front?

Reply:
Great question. Worth a new issue indeed I think.

> But what if the implementation defers dtype casting to numpy (which is currently possible)

Deprecated though, right? The signature should already have included the `dtype` keyword, even though it was still working to not have that.

> then it might be a waste to already copy up front?

What we want to achieve I think is that only a single copy is ever made. We can have two now in many cases, because the other library may copy from its internal implementation into an array, and then numpy will make another copy. The only way to get it right is for the other library to do the copy if `copy=True`, and for numpy not to do so if the method call including `copy=True` succeeded.

Reply:
Created https://github.com/numpy/numpy/issues/25941, in case you want to move over your comments as well